### ISSUE TITLE
Utility analyzers execution time is too long (performance problem)

### ISSUE THREAD

**Issue by duncanp-sonar** [SonarSource/sonar-dotnet#2429]
### Description

See this community thread:
https://community.sonarsource.com/t/help-figuring-out-net-project-scan-taking-significantly-longer/9202/26

To summarise, the `TokenTypeAnalyzer` and `SymbolReferenceAnalyzer` took nearly eight minutes each to run against a specific test project of around 10k lines of code:
```
448.313 50 SonarAnalyzer.Rules.CSharp.TokenTypeAnalyzer
445.228 49 SonarAnalyzer.Rules.CSharp.SymbolReferenceAnalyzer
```

The time taken is excessive for what isn't a large number of lines of code.

Note: the project being analyzed was a test project so only a subset of the utility analyzers are being run. It's possible that the same issue exists with all of the utility analyzers.

### Known workarounds
Exclude the project from analysis by adding the `SonarQubeExclude` property:

```
  <PropertyGroup>
    <SonarQubeExclude>true</SonarQubeExclude> 
  </PropertyGroup>
```

### Related information

* SonarC# Version: v7.13

**Comment by andrei-epure-sonarsource** [comment]
Might help: https://benchmarkdotnet.org/

**Comment by andrei-epure-sonarsource** [comment]
another [community ticket](https://community.sonarsource.com/t/sonarscanner-msbuild-sonar-showprofiling-not-working/15582/3)

**Comment by andrei-epure-sonarsource** [comment]
[This community ticket](https://community.sonarsource.com/t/net-analysis-time-during-msbuild-increased-from-15-min-to-40-min/32898/8) has the same problem

```
33%      SonarAnalyzer.Rules.CSharp.TokenTypeAnalyzer
22%      SonarAnalyzer.Rules.CSharp.SymbolReferenceAnalyzer
18%      SonarAnalyzer.Rules.CSharp.MetricsAnalyzer
12%      SonarAnalyzer.Rules.CSharp.CookieShouldBeHttpOnly
8%      SonarAnalyzer.Rules.CSharp.CopyPasteTokenAnalyzer
````

**Comment by andrei-epure-sonarsource** [comment]
This has been raised in another community topic ([public thread](https://community.sonarsource.com/t/azure-dev-ops-c-slow-build-times/33007) , [private thread](https://community.sonarsource.com/t/re-azure-dev-ops-c-slow-build-times/33108)).

The user sent us a reproducer which we can use to find the perf hotspot in our code.

The file has 5800 LOC which instantiate a huge dictionary, and our analysis is put to its knees due to a single dictionary instantiation.

```
  public class FooCollection<K, V> : Dictionary<K, KeyValuePair<int, FooCollection<K, V>>>
  {
    // stuff
  };

public static class Bar
  {
    public static readonly ReadOnlyDictionary<string, KeyValuePair<int, FooCollection<string, string>>> Foo = new ReadOnlyDictionary<string, KeyValuePair<int, FooCollection<string, string>>>(new FooCollection<string, string>()
    { 
   // lots and lots of nested initializations 
    }; // end of 5800 LOC
```

**Comment by andrei-epure-sonarsource** [comment]
I gave the wrong issue number in my comment above. I meant SONARSEC-1471, which got then ported in sonar-dotnet in #4057.

We have the [IsToComplex()](https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.CFG/ControlFlowGraph/CSharpControlFlowGraphBuilder.cs#L500) method, and we don't generate CFGs for complex methods.

We should do the same to avoid the edge case of large files which literally block the build (takes 2h to analyze).

This won't fix the bigger issue of poor performance of token and symbol generation, but that's normally tolerable.

**Comment by costin-zaharia-sonarsource** [comment]
I've added a reproducer here: https://github.com/costin-zaharia-sonarsource/repro-2429

Most of the time is spent while retrieving the symbols. See [here](https://github.com/SonarSource/sonar-dotnet/blob/814b2c15a10bd83e501d3758b9625f9b1ed3cafc/analyzers/src/SonarAnalyzer.Common/Rules/Utilities/TokenTypeAnalyzerBase.cs#L155) and [here](https://github.com/SonarSource/sonar-dotnet/blob/814b2c15a10bd83e501d3758b9625f9b1ed3cafc/analyzers/src/SonarAnalyzer.Common/Rules/Utilities/SymbolReferenceAnalyzerBase.cs#L133).

**Comment by costin-zaharia-sonarsource** [comment]
Possible solutions:
- cache the retrieved symbols (https://github.com/SonarSource/sonar-dotnet/issues/4218). Both `TokenTypeAnalyzer` and `SymbolReferenceAnalyzer` retrieve the same symbols. By caching them we will reduce the processor time used by these 2 rules in half. I did a POC and for the repro I've added I have the following results:

from:
```
    31.131   28      SonarAnalyzer.Rules.CSharp.TokenTypeAnalyzer
    31.065   28      SonarAnalyzer.Rules.CSharp.SymbolReferenceAnalyzer
```
to: 
```
    29.596   41      SonarAnalyzer.Rules.CSharp.TokenTypeAnalyzer
     0.019   <1      SonarAnalyzer.Rules.CSharp.SymbolReferenceAnalyzer
```

- get the symbols only for those identifiers which appear multiple times in file (SonarQube has support only for in file highlighting) (https://github.com/SonarSource/sonar-dotnet/issues/4220)

- refactor the utility analyzers to execute common code only once (https://github.com/SonarSource/sonar-dotnet/issues/4221)

**Comment by andrei-epure-sonarsource** [comment]
> skip files with a large number of identifiers (we should double check if this step is needed after implementing the first 3)

Time-wise, I suggest to implement the easy-fix and ship it. Implementing the first three suggestions is the ideal scenario, but we want to alleviate the pain of our customers who fall under this case as soon as possible, now that we know this root cause.

**Comment by andrei-epure-sonarsource** [comment]
@costin-zaharia-sonarsource given the time-frame, I have a strong-preference for excluding large/complex files from this.

**Comment by costin-zaharia-sonarsource** [comment]
> given the time-frame, I have a strong-preference for excluding large/complex files from this.

We can create a new ticket and add it to the hardening sprint we will do next week.

In any case we should check also the overhead we add by navigating the tree to get the depth (to check if the syntax tree is to complex). I expect to be insignificant but anyway, we should check.

**Comment by andrei-epure-sonarsource** [comment]
I opened #4273. I am taking this off the kanban because properly fixing the general scenario is not feasible in a couple of days. Let me know if you disagree, @costin-zaharia-sonarsource .

**Comment by pavel-mikula-sonarsource** [comment]
#4217 and #4220 should address this issue
